File: apps/backend/src/idempotency-store.js
const DEFAULT_TTL_MS = 5 * 60 * 1000;
const DEFAULT_MAX_ENTRIES = 2000;

export function createIdempotencyStore({ ttlMs = DEFAULT_TTL_MS, maxEntries = DEFAULT_MAX_ENTRIES } = {}) {
  const cache = new Map();

  function cleanup() {
    const now = Date.now();
    for (const [key, entry] of cache.entries()) {
      if (entry.expiresAt <= now) {
        cache.delete(key);
      }
    }

    if (cache.size <= maxEntries) return;

    const sorted = Array.from(cache.entries()).sort((a, b) => a[1].createdAt - b[1].createdAt);
    const overflow = cache.size - maxEntries;
    for (let i = 0; i < overflow; i += 1) {
      cache.delete(sorted[i][0]);
    }
  }

  return {
    get(key, fingerprint) {
      const entry = cache.get(key);
      if (!entry) return { status: 'miss' };
      if (entry.expiresAt <= Date.now()) {
        cache.delete(key);
        return { status: 'miss' };
      }

      if (fingerprint && entry.fingerprint !== fingerprint) {
        return { status: 'conflict' };
      }

      return { status: 'hit', value: entry.value };
    },
    set(key, fingerprint, value) {
      cleanup();
      cache.set(key, {
        fingerprint,
        value,
        createdAt: Date.now(),
        expiresAt: Date.now() + ttlMs,
      });
    },
  };
}

File: apps/backend-java/src/main/java/uz/reestrmkd/backend/application/IdempotencyStore.java
package uz.reestrmkd.backend.application;

import org.springframework.stereotype.Component;
import uz.reestrmkd.backend.config.AppProperties;

import java.util.Comparator;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Component
public class IdempotencyStore {
    public record Entry(String fingerprint, Object value, long createdAt, long expiresAt) {}
    private final Map<String, Entry> cache = new ConcurrentHashMap<>();
    private final long ttlMs;
    private final int maxEntries;

    public IdempotencyStore(AppProperties props) {
        this.ttlMs = props.getIdempotency().getTtlMs();
        this.maxEntries = props.getIdempotency().getMaxEntries();
    }

    public synchronized Entry get(String key) {
        Entry e = cache.get(key);
        if (e == null) return null;
        if (e.expiresAt() <= System.currentTimeMillis()) {
            cache.remove(key);
            return null;
        }
        return e;
    }

    public synchronized void set(String key, String fingerprint, Object value) {
        cleanup();
        long now = System.currentTimeMillis();
        cache.put(key, new Entry(fingerprint, value, now, now + ttlMs));
    }

    private void cleanup() {
        long now = System.currentTimeMillis();
        cache.entrySet().removeIf(e -> e.getValue().expiresAt() <= now);
        if (cache.size() <= maxEntries) return;
        cache.entrySet().stream()
            .sorted(Comparator.comparingLong(v -> v.getValue().createdAt()))
            .limit(cache.size() - maxEntries)
            .map(Map.Entry::getKey)
            .toList()
            .forEach(cache::remove);
    }
}
